js/core/utils/shadow_dom.js

```diff
@@ -60,3 +60,34 @@ export function addShadowDomStyles($element) {
 
     root.adoptedStyleSheets = [constructedStyleSheet];
 }
+
+function isPositionInElementRectangle(element, x, y) {
+    const rect = element.getBoundingClientRect();
+
+    return rect && x >= rect.left && x < rect.right && y >= rect.top && y < rect.bottom;
+}
+
+export function getShadowElementsFromPoint(x, y, root) {
+    const result = [];
+    const elementQueue = [root];
+
+    while(elementQueue.length) {
+        const el = elementQueue.shift();
```

Comment:
This code gives us `O(n^2)` asymptotic. If we need to implement BFS it's good idea to implement queue data structure (using Object foe example)

Reply:
ok